src/kakarot/instructions/memory_operations.cairo

```diff
@@ -53,4 +55,27 @@ namespace MemoryOperations {
         let ctx = ExecutionContext.increment_gas_used(ctx, GAS_COST_MSTORE);
         return ctx;
     }
+
+    // @notice PC operation
+    // @dev Get the value of the program counter prior to the increment.
+    // @custom:since Frontier
+    // @custom:group Stack Memory Storage and Flow operations.
+    // @custom:stack_produced_elements 1
+    // @return Updated execution context.
+    func exec_pc{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        ctx: model.ExecutionContext*
+    ) -> model.ExecutionContext* {
+        alloc_locals;
+        %{ print("0x58 - PC") %}
+
```

Comment:
We should return PC prior to the execution of PC opcode. With this implementation the returned PC is not correct because we already incremented it after reading PC opcode. 
You can check on https://evm.codes on the playground.
Please add also a test in test_basic.Cairo and associated JSON file.
Also please update the readme with the number of the supported opcodes.

Reply:
Thank you for the review!

Two questions:

1. I believe I understand the functionality after reviewing [https://evm.codes/](https://evm.codes/) but I'm slightly confused as to where the increment is occurring in the `exec_pc()` function. To me, the function appears to read in the current PC of the execution context -> push the value of the PC to the stack -> update the execution context with the current stack -> increment the gas used. So I would assume that if `pc` is initially set to 0, the returned `ctx` still has `pc` set to 0 as I did not explicitly increment `pc` in the code. I understand that after `exec_pc()` runs, the PC should update but where in the code does that occur?
2. How does the test JSON file convention work? I understand everything aside from the value in the `code` field.

Thank you for the guidance!

Reply:
1. It occurs here: https://github.com/sayajin-labs/kakarot/blob/ce5c178a4d4b5f8efe8933281743085d2d1b3e99/src/kakarot/instructions.cairo#L76
2. The code field is the EVM bytecode to run for the test. Let me know if this is not clear

Reply:
1. Oh I believe I get it now, so it's as simple as setting `pc` to `ctx.program_counter - 1`? 
2. Okay understood and how would we generate the bytecode for our implementation? I saw you mentioned it's possible in the evm.codes playground but I'm not 100% sure on how that would work.

Reply:
1. In normal case yes. But once we have JUMP it won’t work. We will likely need to store the last value of the PC, and update it after each instruction 
2. For simple bytecode you can just create it manually.